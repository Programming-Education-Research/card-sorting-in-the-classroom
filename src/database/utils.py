from collections.abc import Generator
from contextlib import contextmanager
from sqlite3 import Cursor, Row, Connection


def dict_factory(cur: Cursor, row: Row):
    fields = [column[0] for column in cur.description]
    return {key: value for key, value in zip(fields, row)}


@contextmanager
def cursor(con: Connection) -> Generator[Cursor]:
    cur = con.cursor()
    try:
        cur.row_factory = dict_factory
        yield cur
    finally:
        cur.close()
