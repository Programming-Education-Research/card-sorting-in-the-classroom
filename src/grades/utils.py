from grades.grade import Grade


def from_raw(data: dict[str, str | float]):
    return Grade(
        username=data["username"],
        grade=calculate_average(data),
        sat_exam="Exam" in data,
    )


def calculate_average(data):
    marks = [v for k, v in data.items() if k.startswith("Test") or k == "Exam"]
    return sum(marks) / len(marks) if marks else 0
