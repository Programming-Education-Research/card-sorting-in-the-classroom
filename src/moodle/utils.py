from collections import defaultdict


def join_by(key, *args):
    key_to_result = defaultdict(dict)
    for col in args:
        for item in col:
            key_to_result[item[key]] |= item
    return list(key_to_result.values())