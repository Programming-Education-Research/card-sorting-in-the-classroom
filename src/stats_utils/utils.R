library(tidyverse)
library(DBI)
library(knitr)

question_ordering <- c(
    "Math Expressions",
    "Boolean Reasoning",
    "Dicts",
    "Dicts With Errors",
    "Functional Equivalence",
    "Identify Exceptions",
    "Test Case Match"
)

relabel_courses <- function(course) ordered(
    case_match(course, "CS101" ~ "Compsci 101", "CS130" ~ "Compsci 130"),
    levels = c("Compsci 101", "Compsci 130")
)

relabel_semester <- function(semester) ordered(
    case_match(semester, "SS2025" ~ "Summer Semester", "S12025" ~ "Semester One"),
    levels = c("Summer Semester", "Semester One")
)

relabel_exam <- function(exam) ordered(
    case_match(exam, 1 ~ "Sat Exam", 0 ~ "Did Not Sit"),
    levels = c("Sat Exam", "Did Not Sit")
)

relabel_questions <- function(questions) ordered(
    questions,
    levels = question_ordering
)

t <- function(table) {
    if (isTRUE(getOption('knitr.in.progress'))) {
        return(knitr::kable(table))
    } else {
        return(table)
    }
}


fetch <- function(query) {
    mydb <- dbConnect(RSQLite::SQLite(), "data/database.db")
    raw_data <- dbGetQuery(mydb, query)
    dbDisconnect(mydb)
    return(raw_data)
}


custom_theme <- function(...) (
    theme_minimal()
    + theme(
        plot.title = element_text(face = "bold"),
        axis.text = element_text(colour = "black", face = "bold", size = 10),
        axis.title = element_text(colour = "black", face = "bold", size = 11),
        panel.grid = element_line(
            colour = "black", linetype = "dotted", linewidth = 0.3),
        panel.border = element_rect(linetype = "solid", fill = NA),
        strip.text = element_text(colour = "black", face = "bold", size = 12),
        plot.caption.position = "plot",
        plot.background = element_rect(fill = "white", colour = "white"),
    )
    + theme(...)
)

is_outlier <- function(values) {
    return(values > quantile(values, 0.75) + 1.5 * IQR(values, na.rm = TRUE))
}
