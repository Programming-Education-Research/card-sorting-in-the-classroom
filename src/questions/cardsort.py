import json
from dataclasses import dataclass
from typing import Self, override, Final

from database.orm import field
from moodle.xml import RawQuestion
from questions.question import Question

CARD_SPEC: Final = {
    "Cards": {
        "type": "array",
        "max_items": 0,
        "items": {"type": "string"},
        "strict": True
    }
}


@dataclass
class Cardsort(Question):
    prompt: str
    preload: field(
        dict[str, list[str]],
        marshal=json.dumps,
        unmarshal=json.loads,
    )
    answer: field(
        dict[str, list[str]],
        marshal=json.dumps,
        unmarshal=json.loads,
    )
    is_ordered: bool

    @classmethod
    @override
    def from_raw(cls, moodle_xml: RawQuestion) -> Self:
        return Cardsort(
            name=moodle_xml["name"],
            prompt=moodle_xml["prompt"],
            preload=parse_groups(moodle_xml["preload"]),
            answer=parse_groups(moodle_xml["answer"]),
            is_ordered=parse_ordering(moodle_xml["preload"]),
        )

    @override
    def json_to_attempt(
          self,
          json_object: dict[str, list[str]]
    ) -> dict[str, list[str]]:
        labelled_cards = {
            f"Card {i}": card
            for i, card in enumerate(self.cards(), start=1)
        }
        return {
            group: [labelled_cards[card] for card in cards]
            for group, cards in json_object.items()
        }

    @override
    def json_schema(self) -> dict[str, object]:
        labels = [f"Card {i}" for i in range(1, len(self.cards()) + 1)]
        groups = {
            group: {
                "type": "array",
                "items": {"type": "string", "enum": labels},
                "strict": True,
            }
            for group in self.preload
            if group.lower() != "cards"
        }
        return {
            "type": "object",
            "properties": groups | CARD_SPEC,
            "required": ["Cards", *groups],
            "additionalProperties": False,
            "strict": True,
        }

    @override
    def instructions(self) -> str:
        return self.prompt

    @override
    def question(self) -> str:
        labelled_cards = {
            card: f"Card {i}\n{card}"
            for i, card in enumerate(self.cards(), start=1)
        }
        groups = {
            group: [labelled_cards[card] for card in cards]
            for group, cards in self.preload.items()
        }
        return json.dumps(groups)

    def cards(self) -> list[str]:
        return [card for group in self.preload.values() for card in group]


def parse_ordering(preload: str) -> bool:
    data = json.loads(preload)
    return data["cardsort"][0]["config"]["groupOrder"]


def parse_groups(data: str) -> dict[str, list[str]]:
    data = json.loads(data)
    groups = data["cardsort"][0]["groups"]
    return {
        group["title"]: [card["prompt"] for card in group["cards"]]
        for group in groups
    }
