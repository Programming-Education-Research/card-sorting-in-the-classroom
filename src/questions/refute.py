import re
from dataclasses import dataclass
from typing import Self, override

from moodle.xml import RawQuestion
from questions.question import Question


@dataclass
class Refute(Question):
    prompt: str
    preload: str
    test: str

    @classmethod
    @override
    def from_raw(cls, moodle_xml: RawQuestion) -> Self:
        return Refute(
            name=moodle_xml["name"],
            prompt=moodle_xml["prompt"],
            preload=moodle_xml["globalextra"],
            test=moodle_xml["tests"][0]["code"],
        )

    @override
    def json_to_attempt(self, json_object: object) -> object:
        return json_object

    @override
    def json_schema(self) -> dict[str, object]:
        return {
            "type": "object",
            "properties": {
                "given": {"type": "string", "strict": True, },
                "then": {"type": "string", "strict": True, },
                "but": {"type": "string", "strict": True, },
            },
            "required": ["given", "then", "but"],
            "strict": True,
        }

    @override
    def instructions(self) -> str:
        return self.prompt

    @override
    def question(self) -> str:
        parts = re.split(r"\{\s*\[\s*\d+\s*\]\s*\}", self.preload)
        return "???".join(parts)
