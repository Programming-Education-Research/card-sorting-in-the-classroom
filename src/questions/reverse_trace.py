import re
from dataclasses import dataclass
from typing import Self, override

from moodle.xml import RawQuestion
from questions.question import Question


@dataclass
class ReverseTrace(Question):
    prompt: str
    preload: str
    expect: str

    @classmethod
    @override
    def from_raw(cls, moodle_xml: RawQuestion) -> Self:
        return ReverseTrace(
            name=moodle_xml["name"],
            prompt=moodle_xml["prompt"],
            preload=moodle_xml["globalextra"],
            expect=moodle_xml["tests"][0]["stdout"],
        )

    @override
    def json_to_attempt(self, json_object: object) -> object:
        return json_object

    @override
    def json_schema(self) -> dict[str, object]:
        return {"type": "string", "strict": True}

    @override
    def instructions(self) -> str:
        return self.prompt

    @override
    def question(self) -> str:
        return re.sub(r"\{\s*\[\s*\d+\s*\]\s*\}", "???", self.preload)
