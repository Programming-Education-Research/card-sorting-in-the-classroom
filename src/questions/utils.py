from moodle.xml import RawQuestion
from questions.cardsort import Cardsort
from questions.question import Question
from questions.refute import Refute
from questions.reverse_trace import ReverseTrace


def from_raw(raw: RawQuestion) -> Question:
    match raw["type"]:
        case "python_refute":
            return Refute.from_raw(raw)
        # I have no idea why, but the production Moodle instance suffixes a 1
        case "python3_html_gapfiller-1":
            return ReverseTrace.from_raw(raw)
        case "cardsort":
            return Cardsort.from_raw(raw)
        case _:
            raise RuntimeError(f"Unrecognised question type: `{raw['type']}`")
