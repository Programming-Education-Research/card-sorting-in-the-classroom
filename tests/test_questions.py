from approvaltests import verify

from questions.cardsort import Cardsort
from questions.refute import Refute
from questions.reverse_trace import ReverseTrace
from utils import test

A_REFUTE_QUESTION = Refute(
    "The refute question name",
    "The refute question prompt",
    "Given: foo({[12]}); Then: {[12]}; But: {[12]}",
    "Given: {[given]}; Then: {[then]}; But: {[but]}",
)
A_REVERSE_TRACE_QUESTION = ReverseTrace(
    "The revere trace question name",
    "The reverse trace question prompt",
    "foo = {[12]}",
    "12345",
)
A_CARD_SORT_QUESTION = Cardsort(
    "The card sort question name",
    "The card sort question prompt",
    {"Foo": ["A", "B", "C"], "Bar": [], "Baz": []},
    {"Foo": [], "Bar": ["A", "B"], "Baz": ["C"]},
    False,
)


@test
def refute_questions_give_instructions():
    verify(A_REFUTE_QUESTION.instructions())


@test
def refute_questions_give_questions():
    verify(A_REFUTE_QUESTION.question())


@test
def refute_questions_parse_completions():
    answer = {"given": "foo", "then": "bar", "but": "baz"}
    assert A_REFUTE_QUESTION.json_to_attempt(answer) == answer


@test
def reverse_trace_questions_give_instructions():
    verify(A_REVERSE_TRACE_QUESTION.instructions())


@test
def reverse_trace_questions_give_questions():
    verify(A_REVERSE_TRACE_QUESTION.question())


@test
def reverse_trace_questions_parse_completions():
    answer = "Foo"
    assert A_REVERSE_TRACE_QUESTION.json_to_attempt(answer) == answer

@test
def card_sort_questions_give_instructions():
    verify(A_CARD_SORT_QUESTION.instructions())


@test
def card_sort_questions_give_questions():
    verify(A_CARD_SORT_QUESTION.question())

@test
def card_sort_questions_parse_completions():
    answer = {"Foo": [], "Bar": ["Card 1", "Card 2"], "Baz": ["Card 3"]}
    expected = {"Foo": [], "Bar": ["A", "B"], "Baz": ["C"]}
    assert A_CARD_SORT_QUESTION.json_to_attempt(answer) == expected
