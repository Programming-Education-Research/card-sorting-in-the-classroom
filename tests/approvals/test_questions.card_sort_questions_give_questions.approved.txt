{"Foo": ["Card 1\nA", "Card 2\nB", "Card 3\nC"], "Bar": [], "Baz": []}
